packages/camera/camera_windows/windows/capture_controller.h

```diff
@@ -0,0 +1,285 @@
+
+  // Handlers for CaptureEngineListener events
+  // From CaptureEngineObserver
+  bool IsReadyForSample() override {
+    return initialized_ && previewing_ && !preview_paused_;
+  }
+
+  void OnEvent(IMFMediaEvent* event) override;
+
+  uint8_t* GetSourceBuffer(uint32_t current_length) override;
+  void OnBufferUpdated() override;
+  void UpdateCaptureTime(uint64_t capture_time) override;
+
+  // Sets capture engine, for mocking purposes
+  void SetCaptureEngine(IMFCaptureEngine* capture_engine) {
+    capture_engine_ = capture_engine;
+  };
+
+  // Sets video source, for mocking purposes
+  void SetVideoSource(IMFMediaSource* video_source) {
+    video_source_ = video_source;
+  };
+
+  // Sets audio source, for mocking purposes
+  void SetAudioSource(IMFMediaSource* audio_source) {
+    audio_source_ = audio_source;
+  };
+
+ private:
+  CaptureControllerListener* capture_controller_listener_ = nullptr;
+  bool initialized_ = false;
+  bool enable_audio_record_ = false;
+  std::string video_device_id_;
+
+  ResolutionPreset resolution_preset_ =
+      ResolutionPreset::RESOLUTION_PRESET_MEDIUM;
+
+  // CaptureEngine objects
+  bool capture_engine_initialization_pending_ = false;
+  ComPtr<IMFCaptureEngine> capture_engine_;
+  ComPtr<CaptureEngineListener> capture_engine_callback_handler_;
+
+  ComPtr<IMFDXGIDeviceManager> dxgi_device_manager_;
+  ComPtr<ID3D11Device> dx11_device_;
+  // ID3D12Device* dx12_device_ = nullptr;
+  UINT dx_device_reset_token_ = 0;
+
+  // Sources
+  ComPtr<IMFMediaSource> video_source_;
+  ComPtr<IMFMediaSource> audio_source_;
+
+  // Texture
+  int64_t texture_id_ = -1;
+  flutter::TextureRegistrar* texture_registrar_ = nullptr;
+  std::unique_ptr<flutter::TextureVariant> texture_;
+
+  // TODO: add release_callback and clear buffer after each frame
+  FlutterDesktopPixelBuffer flutter_desktop_pixel_buffer_ = {};
+  uint32_t source_buffer_size_ = 0;
+  std::unique_ptr<uint8_t[]> source_buffer_data_ = nullptr;
+  std::unique_ptr<uint8_t[]> dest_buffer_ = nullptr;
+  uint32_t bytes_per_pixel_ = 4;  // MFVideoFormat_RGB32
+
+  // Preview
+  bool preview_paused_ = false;
+  bool preview_pending_ = false;
+  bool previewing_ = false;
+  uint32_t preview_frame_width_ = 0;
+  uint32_t preview_frame_height_ = 0;
+  ComPtr<IMFMediaType> base_preview_media_type_;
+  ComPtr<IMFCapturePreviewSink> preview_sink_;
+
+  // Photo / Record
+  bool pending_image_capture_ = false;
+  bool record_start_pending_ = false;
+  bool record_stop_pending_ = false;
+  bool recording_ = false;
+  int64_t record_start_timestamp_us_ = -1;
+  uint64_t recording_duration_us_ = 0;
+  int64_t max_video_duration_ms_ = -1;
+
+  uint32_t capture_frame_width_ = 0;
+  uint32_t capture_frame_height_ = 0;
+  ComPtr<IMFMediaType> base_capture_media_type_;
+  ComPtr<IMFCapturePhotoSink> photo_sink_;
+  ComPtr<IMFCaptureRecordSink> record_sink_;
+  std::string pending_picture_path_ = "";
+  std::string pending_record_path_ = "";
```

Comment:
This is a *lot* of ivars for one class. If you have this many things, and they have their own groupings and headings, that's a strong signal that you need to break out responsibilities into helper classes.

Reply:
Yes, I'll refactor these. Some of the logic can be moved to helper classes, like texture handling.
Some of the states can be done with enum states instead of having multiple booleans, to be more error tolerable as well.
This has grown on development and definitely needs some refactoring.